### PR TITLE
tls/sni: refactored to reduce dependencies and improve Dispatcher.Shutdown()

### PR DESCRIPTION
**Pull request by amery**

### CODE REVIEW

File: tls/sni/dispatcher.go [hunk 1]
```diff
@@ -225,12 +225,23 @@ func (d *Dispatcher) onError(err error) error {
 	return nil
 }
 
-// Shutdown initiates a shutdown and waits until the workers are done.
-// The given Context isn't used, its only to allow implementing the
-// httpgroup.Server interface
-func (d *Dispatcher) Shutdown(context.Context) error {
+// Shutdown initiates a shutdown and waits until the workers are done
+// or the given context to timeout.
```
Comment (karasz) [inline tls/sni/dispatcher.go:229]: or the given context times out.

**Review by karasz — CHANGES_REQUESTED** [review]

**Comment by amery** [comment]
amended as requested

**Review by karasz — APPROVED** [review]
LGTM

**Comment by amery** [comment]
`darvaza.org/x/tls@v0.1.2` tagged

**Pull request merged**